# 06 EDA — Reviews

Read-only inspection of `olist_order_reviews_dataset.csv`. No cleaning, charts, or feature engineering.

## 1. Import libraries

Import pandas for tabular inspection.

In [1]:
import pandas as pd

**Observation:** pandas is ready.

## 2. Load CSV

Load only this CSV into memory. Timestamp columns are converted on the in-memory copy.

In [2]:
file_path = '../data/raw/olist/olist_order_reviews_dataset.csv'
df = pd.read_csv(file_path)
for column in ['review_creation_date', 'review_answer_timestamp']:
    df[column] = pd.to_datetime(df[column], errors='coerce')

**Observation:** The table is loaded; the raw CSV is unchanged.

## 3. Shape

Check table size.

In [3]:
df.shape

(99224, 7)

**Observation:** This reports the table's row and column count.

## 4. First 5 rows

Preview the raw structure.

In [4]:
df.head()

                          review_id  ... review_answer_timestamp
0  7bc2406110b926393aa56f80a40eba40  ...     2018-01-18 21:46:59
1  80e641a11e56f04c1ad469d5645fdfde  ...     2018-03-11 03:05:13
2  228ce5500dc1d8e020d8d1322874b6f0  ...     2018-02-18 14:36:24
3  e64fb393e7b32834bb789ff8bb30750e  ...     2017-04-21 22:02:06
4  f7c4243c7fe1938f181bec41a392bdeb  ...     2018-03-02 10:26:53

[5 rows x 7 columns]

**Observation:** The preview confirms the row grain and field layout.

## 5. Column names

List available fields.

In [5]:
df.columns.tolist()

['review_id', 'order_id', 'review_score', 'review_comment_title', 'review_comment_message', 'review_creation_date', 'review_answer_timestamp']

**Observation:** These are the fields available for this table's analysis.

## 6. Data types

Review pandas data types.

In [6]:
df.dtypes

review_id                             str
order_id                              str
review_score                        int64
review_comment_title                  str
review_comment_message                str
review_creation_date       datetime64[us]
review_answer_timestamp    datetime64[us]
dtype: object

**Observation:** Types show which fields are text, numeric, or converted datetime values.

## 7. Missing values

Count nulls by field.

In [7]:
df.isna().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

**Observation:** Missing counts identify incomplete fields without changing them.

## 8. Missing percentage

Measure missingness relative to table size.

In [8]:
(df.isna().mean().mul(100).round(2)).sort_values(ascending=False)

review_comment_title       88.34
review_comment_message     58.70
review_id                   0.00
review_score                0.00
order_id                    0.00
review_creation_date        0.00
review_answer_timestamp     0.00
dtype: float64

**Observation:** Percentages show the relative scale of incomplete fields.

## 9. Duplicate rows

Count complete row duplicates.

In [9]:
df.duplicated().sum()

np.int64(0)

**Observation:** This is the number of repeated full rows; none are removed.

## 10. Duplicate primary keys

Validate the documented candidate key.

In [10]:
primary_key = ['review_id', 'order_id']
df.duplicated(subset=primary_key, keep=False).sum()

np.int64(0)

**Observation:** This counts rows involved in duplicate candidate keys for review_id, order_id.

## 11. Unique values

Inspect meaningful category or entity counts.

In [11]:
df['review_score'].value_counts(dropna=False).sort_index()

review_score
1    11424
2     3151
3     8179
4    19142
5    57328
Name: count, dtype: int64

**Observation:** This gives a concise view of important distinct values.

## 12. Basic describe()

Generate a basic summary for all fields.

In [12]:
df.describe(include='all')

                               review_id  ...     review_answer_timestamp
count                              99224  ...                       99224
unique                             98410  ...                         NaN
top     c444278834184f72b1484dfe47de7f97  ...                         NaN
freq                                   3  ...                         NaN
mean                                 NaN  ...  2018-01-16 00:23:56.977939
min                                  NaN  ...         2016-10-07 18:32:28
25%                                  NaN  ...  2017-09-27 01:53:27.250000
50%                                  NaN  ...  2018-02-04 22:41:47.500000
75%                                  NaN  ...  2018-05-20 12:11:21.500000
max                                  NaN  ...         2018-10-29 12:27:35
std                                  NaN  ...                         NaN

[11 rows x 7 columns]

**Observation:** The summary describes completeness, distinctness, common values, and numeric ranges where applicable.

## 13. Table-specific observations

Reviews provide order-level ratings and optional customer comments. Scores can later connect fulfilment performance with satisfaction.

## 14. Data quality observations

Review titles are missing in 87,656 rows and messages in 58,247 rows. The composite (`review_id`, `order_id`) is valid, but neither field alone is unique; all reviewed orders exist.